src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs

```diff
@@ -0,0 +1,368 @@
+        {
+            PluginCreationResult pluginCreationResult = null;
+
+            if (result.PluginFile.State == PluginFileState.Valid)
+            {
+                var plugin = await _pluginFactory.GetOrCreateAsync(
+                    result.PluginFile.Path,
+                    PluginConstants.PluginArguments,
+                    new RequestHandlers(),
+                    _connectionOptions,
+                    cancellationToken);
+
+                var utilities = await PerformOneTimePluginInitializationAsync(plugin, cancellationToken);
+
+                plugin.Connection.ProtocolVersion.Equals(Plugins.ProtocolConstants.CurrentVersion);
+
+                var lazyOperationClaims = _pluginOperationClaims.GetOrAdd(
+                        requestKey,
+                        key => new Lazy<Task<IReadOnlyList<OperationClaim>>>(() =>
+                        GetPluginOperationClaimsAsync(
+                            plugin,
+                            packageSourceRepository,
+                            serviceIndex,
+                            cancellationToken)));
+
+                await lazyOperationClaims.Value;
+
+                pluginCreationResult = new PluginCreationResult(
+                    plugin,
+                    utilities.Value,
+                    lazyOperationClaims.Value.Result);
+            }
+            else
+            {
+                pluginCreationResult = new PluginCreationResult(result.Message);
+            }
+
+            return pluginCreationResult;
+        }
+
+
+        private async Task<Lazy<IPluginMulticlientUtilities>> PerformOneTimePluginInitializationAsync(IPlugin plugin, CancellationToken cancellationToken)
+        {
+            var utilities = _pluginUtilities.GetOrAdd(
+                                plugin.Id,
+                                path => new Lazy<IPluginMulticlientUtilities>(
+                                    () => new PluginMulticlientUtilities()));
+
+            await utilities.Value.DoOncePerPluginLifetimeAsync(
+                MessageMethod.MonitorNuGetProcessExit.ToString(),
+                () => plugin.Connection.SendRequestAndReceiveResponseAsync<MonitorNuGetProcessExitRequest, MonitorNuGetProcessExitResponse>(
+                    MessageMethod.MonitorNuGetProcessExit,
+                    new MonitorNuGetProcessExitRequest(_currentProcessId.Value),
+                    cancellationToken),
+                cancellationToken);
+
+            await utilities.Value.DoOncePerPluginLifetimeAsync(
+                MessageMethod.Initialize.ToString(),
+                () => InitializePluginAsync(plugin, _connectionOptions.RequestTimeout, cancellationToken),
+                cancellationToken);
+            return utilities;
+        }
+
+        private void Reinitialize(IEnvironmentVariableReader reader,
```

Comment:
Rename to `Initialize` since it's only called once per instance.

Reply:
done.